### ISSUE TITLE
Project64.rdb should use 1's and 0's instead of "Yes" or "No"

### ISSUE THREAD

**Issue by oddMLan** [project64/project64#238]
It would be less confusing that way. Also to be consistent with the CFG file.

**Comment by AmbientMalice** [comment]
I think consistency is overall a good idea. Having "Off" "No" and "0" all representing the same thing but not compatible with each other is kind of unnecessarily fiddly.

**Comment by cxd4** [comment]
I remember asking but the reason project64 said it was this way was because the RDB file would retain at least partial backwards-compatibility with previous versions of the RDB.

That being said, the 32bit option wasn't available in Project64 1.6, or 1.7, so if zilmar were to change at least that particular option from "Yes"/"No" to '1'/'0' there would be less backwards compatibility to have broken, since it's new to 2.x.  Still, if people wanted to compile older builds from older versions of the repo, the RDB file would no longer work in that area.

**Comment by oddMLan** [comment]
If backwards compatibility is a concern then I'd suggest to at least make the paramaters compatible with each other, that way we could retain backwards compatibility with older RDBs and potentially have a less confusing and cleaner looking RDB.

**Comment by cxd4** [comment]
The RDB can't be changed without breaking backwards compatibility.
Making it say "No" instead of "0" or whatever has to work as it does now, the same way as it has since 1.6.  Otherwise yes like you said breaking backward-compat would be a concern.

**Comment by project64** [comment]
@oddMLan it is possible with out to much work to be able to change the cfg file to work the same as the rdb with yes and no instead of 0 and 1

**Comment by cxd4** [comment]
Or that doesn't sound so bad I guess. ^

**Comment by oddMLan** [comment]
>@oddMLan it is possible with out to much work to be able to change the cfg file to work the same as the rdb with yes and no instead of 0 and 1

TBH I prefer 0 and 1 but the cfg and rdb files would at least be consistent now without breaking backwards compability so overall yeah it doesn't sound too bad I guess.

**Comment by Frank-74** [comment]
Why not make 1/Yes/On=1 and 0/No/Off=0, so it would be compatible with both versions.

**Comment by oddMLan** [comment]
>Why not make 1/Yes/On=1 and 0/No/Off=0, so it would be compatible with both versions.
Because as cxd4 said the newer RDBs would not be compatible with older builds.

**Comment by Frank-74** [comment]
I suppose it would be better for both .cfg and .rdb to use the same types.  Which reminds me, the RDB Edit Mode needs some fixing up.  It writes a redundant value on a new line at the end of the game you edit.

**Comment by cxd4** [comment]
> And I think that would not be worth the effort to implement if it's not to change the RDB to use only 0 and 1.

Except that in theory, it reduces the breakage of backwards-compat if the initiative were seized as early as possible to make pj64 have the same result for "0" == "No" = "Off", so that if it were done now rather than later or never at all there would be fewer past versions of pj64 it would break.

But, that runs the risk of just leaving this flexible support sitting around in the EXE as idle and unused code if the RDB was indeed to never be changed.

Basically "0" and "1" is better because it's just (char & 1) or (char - '0') and you have the 0 or 1 integer telling you whether or not a setting is off or on, which is just way smaller and more efficient parsing on a systems level (unlike that stupid `bool` type C++ has, which is mostly only good for packing or sizability).  So using 0/1 instead of No/Yes would be better file size AND performance, but I think, ultimately the perfect emulator would not even be needing a RDB file in the first place...so I think it's hardly that much of a thing to worry about.

**Comment by cxd4** [comment]
Actually, that reminds me of an old tactic I used.

"Yes" and "No":  the 'y' & 1 is 1, 'n' & 1 is 0
"0" and "1":  the '1' & 1 is 1, '0' & 1 is 0

So in a way you could have less code size while creating the room for the future of having either one work.

**Comment by vgturtle127** [comment]
Why is backwards compatibility an issue? Newer versions of Project64 are suppose to be better. If we keep maintaining backwards compatibility so that we don't break older versions, isn't that setting us up for making a lackluster experience for everybody else?

In my opinion, we should just change it now for the better, and if somebody wants to use a new RDB with an old version of Project64, they should not receive support for that, but they should be able to convert it since it is mostly text. Right?

If Project64 is going to get better we are going to have to eventually change a few things. I think the RDB is probably more minor since it is easily editable, and mostly just text with basic values.

Just my opinion, and I'm only saying this because I want Project64 to become easier to use and better overall. :smile: 

In a somewhat related note, we should probably make sure the RDB is somewhat complete before we consider changing the entire thing. Just a thought.

Not trying to push anybody's buttons or anything, just trying to get better N64 emulation. :confounded: 

Between Project64, Mupen64, CEN64, 1964, and all other emulators and forks, you'd think we would have this down by now. It's still so hard. :disappointed:

**Comment by cxd4** [comment]
Backwards compatibility is either more or less of an issue than other things.

In this case it's more of an issue than whether a config file should change "Yes"/"No" to '0'/'1'.
In emulation accuracy cases well it's less of an issue than fixing bugs and things like that.

Have to weigh the pros with the cons.  We've had to go as far back as Project64 version 1.4 to do regression testing--it's just most convenient when you're able to use the same exact configuration between different builds of pj64 2.x, 1.7, 1.6, 1.4 etc. to help test things and isolate regressions.

**Comment by project64** [comment]
Backward compatibly was never the reason, the reason it was designed that way because there was no way to easy way for people to contribute to the RDB file. People would create there own one and by keeping the rdb consistent allowed project64 to always be able to use it.

Now that everything is running on git I do not expect that there will be any 3rd party rdb files any more the real need for this is no longer there. Really git has only been used for 2.2 so I guess it could have been changed for that version.

I guess I have no issues any more with changing this so the cfg file and the rdb match in file format now.

**Comment by vgturtle127** [comment]
Has anybody ever fuzzed the source? Just curious. That may help with that. Regression testing, I mean.

RDB and config stuff are easy to change, update, etc. as long as they are kept text and not converted to binary files. So as long as that remains the same, it should technically be possible to please everybody.

I'm not sure how other emulators run games (probably by trying to find the right offset and launching from there...unless I'm thinking of NES games since they had headers?), but an RDB seems like it doubles as a list of known games, which I think is good to have too. I can see it both ways, is what I'm saying.

I don't know. :confused:

**Comment by oddMLan** [comment]
>Backward compatibly was never the reason, the reason it was designed that way because there was no way to easy way for people to contribute to the RDB file. *People would create there own one and by keeping the rdb consistent allowed project64 to always be able to use it.*

That's something I never considered, but I see the point.

>Now that everything is running on git I do not expect that there will be any 3rd party rdb files any more the real need for this is no longer there

Agreed. It would be nice to do the change now for the impending release of Project64 2.3.

**Comment by vgturtle127** [comment]
I just re-read some of the discussion here, and I just realized that any changes we made to the RDB could be easily searched and replaced even with Project 1.7 and earlier or something. AN example being, if you want an RDB with compatibility for 1.6, just use any text editors search and replace option and change all the values to be compatible with that. But if you want 2.x support, you can just change it again using the search and replace option. Just a thought, I'm only suggesting it wouldn't be a big deal to convert from one to another. You could probably make a batch script for Windows and an .sh script for Linux easily.

**Comment by oddMLan** [comment]
But @project64 said backwards compatibility wasn't the reason the RDB and CFG are different.
Read my latest post again.

What would be the point of updating an RDB for old Project64 versions? If someone wants to do it for whatever reason, that's on them, not on the current project.

**Comment by vgturtle127** [comment]
@oddMLan Alright.

By the way, you guys have at least 30+ ROMs that aren't on the RDB. What is the best method of adding them to the RDB? What information about the ROM is needed? I know there is some stuff you guys don't have a need to emulate (like the GameShark and Action Replay, for example), but it would still be nice to know what the ROM is instead of having to view the info about it...

**Comment by project64** [comment]
closing old issue, if this is still an issue it can be reopened

**Comment by Squall-Leonhart** [comment]
Lets reopen this, 4.0 would be as good a time as any to break away from legacy crud and make the variables consistent for off/on settings as well as uniformity between the user rdb and global rdb.

**Comment by project64** [comment]
https://github.com/project64/project64/commit/2db5c81af5b8b50aefcdc7d49aa23be4ac8dc68a

this should at least change yes/no to 1 and 0

**Comment by oddMLan** [comment]
Cool
I assume that change will also be carried to the Off/On type as well eventually.

**Comment by project64** [comment]
most likely it should be an easy change, then the rdb should match the user config more closely
